File: morse.c
const int LED = 13;

void setup() {
  pinMode(LED, OUTPUT);
  Serial.begin(9600);
}

// Function to blink morse code
void blinkMorse(String message) {
  // Morse code dictionary
  String morseCode[36] = { ".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---", "-.-", ".-..", "--", "-.", "---", ".--.", "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", "-.--", "--..", ".----", "..---", "...--", "....-", ".....", "-....", "--...", "---..", "----.", "-----", "--..--", ".-.-.-", "..--.." };
  
  for (int i = 0; i < message.length(); i++) {
    char c = message[i];
    if (c >= 'a' && c <= 'z') {
      String code = morseCode[c - 'a'];
      for (int j = 0; j < code.length(); j++) {
        if (code[j] == '.') {
          digitalWrite(LED, HIGH);
          delay(200);
          digitalWrite(LED, LOW);
          delay(200);
        } else if (code[j] == '-') {
          digitalWrite(LED, HIGH);
          delay(600);
          digitalWrite(LED, LOW);
          delay(200);
        }
      }
    }
    // If character is a number
    else if (c >= '0' && c <= '9') {
      String code = morseCode[c - '0' + 26];
      for (int j = 0; j < code.length(); j++) {
        if (code[j] == '.') {
          digitalWrite(LED, HIGH);
          delay(200);
          digitalWrite(LED, LOW);
          delay(200);
        } else if (code[j] == '-') {
          digitalWrite(LED, HIGH);
          delay(600);
          digitalWrite(LED, LOW);
          delay(200);
        }
      }
    }
    // Space between letters
    delay(400);
  }
}

void loop() {
  if (Serial.available() > 0) {
    String message = Serial.readString();
    blinkMorse(message);
  }
}

File: morse.rs
use std::{thread, time};

const LED: u64 = 13;

fn blink_morse(message: &str) {
    // Morse code dictionary
    let morse_code = vec![
        ".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---", "-.-", ".-..", "--",
        "-.", "---", ".--.", "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", "-.--", "--..",
        ".----", "..---", "...--", "....-", ".....", "-....", "--...", "---..", "----.", "-----",
        "--..--", ".-.-.-", "..--..",
    ];

    for c in message.chars() {
        if c.is_ascii_alphabetic() {
            let code = morse_code[(c.to_ascii_lowercase() as u8 - b'a') as usize];
            for s in code.chars() {
                if s == '.' {
                    blink(200);
                } else if s == '-' {
                    blink(600);
                }
            }
        } else if c.is_ascii_digit() {
            let code = morse_code[(c as u8 - b'0') as usize + 26];
            for s in code.chars() {
                if s == '.' {
                    blink(200);
                } else if s == '-' {
                    blink(600);
                }
            }
        }
        // Space between letters
        thread::sleep(time::Duration::from_millis(400));
    }
}

fn blink(duration: u64) {
    println!("{}", duration);
    thread::sleep(time::Duration::from_millis(duration));
}

fn main() {
    let message = "Hello World";
    blink_morse(message);
}
